src/dp.rs

```diff
@@ -0,0 +1,64 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! Differential privacy (DP) primitives.
+use crate::field::FieldElement;
+use crate::vdaf::prg::Seed;
+use std::fmt::Debug;
+
+// TODO(tholop): We could also use pairs of integers, or even
+// floats to store the DiscreteGaussian DP parameters?
+// Using BigInts might make it harder to communicate `DiscreteGaussian` to the Collector.
+use num_rational::BigRational;
+
+/// DP-related errors.
+#[derive(Debug, thiserror::Error)]
+pub enum DpError {
+    /// Error while sampling from a specific noise distribution
+    #[error("sampling error")]
+    SamplingError,
+}
+
```

Comment:
It occurs to me that this API bears a striking resemblance to `prg::CoinToss`. Can we use that instead? The differences are:
* Here we use `seed`, but `CointToss::sample()` takes a `seed_stream`. The latter probably makes more sense.
* Here we have a `len` field, which is `CoinToss::sample()` doesn't have. However, I have been playing around with an IDPF application and noticed that I wanted to be able to provide `CoinToss` a "length hint" so that the length of the output shares can be determined at runtime instead of compile time.

What do you think, @divergentdave?

Reply:
I think the big issue is that `CoinToss` implicitly assumes a uniform distribution. Extending it for DP-relevant distributions would require turning them into parallel universe version of `rand`'s traits.

Taking a `SeedStream` mutable reference sounds good. This opens the door to re-using a PRG within one `add_noise_to_agg_share` call, which could be a significant performance win. (`PrgSha3` setup/derivation is relatively expensive)

For the separate IDPF use case, we could write a helper function or provided method to handle calling `CoinToss::sample()` repeatedly and returning a `Vec<T>` of the desired length. But, since `CoinToss` exists to be a trait bound on `IdpfValue`, we may need to take a broader look at the whole IDPF API, as it sounds like `convert()` may need a way to pass that length when producing values.

Reply:
> I think the big issue is that `CoinToss` implicitly assumes a uniform distribution. Extending it for DP-relevant distributions would require turning them into parallel universe version of `rand`'s traits.

It assumes the output of `SeedStream` is uniform, but so would `Distribution`. But you're definitely on to something: we don't want to be duplicating work that has already been done.

My main goal here is to avoid defining two traits in libprio that are nearly identical. Are there traits in that crate we should be implementing instead? 

> For the separate IDPF use case, we could write a helper function or provided method to handle calling `CoinToss::sample()` repeatedly and returning a `Vec<T>` of the desired length. But, since `CoinToss` exists to be a trait bound on `IdpfValue`, we may need to take a broader look at the whole IDPF API, as it sounds like `convert()` may need a way to pass that length when producing values.

It's not super urgent, but it is something I'd like to investigate at some point. I do think it's going to be useful to have. I'll file an issue.

Reply:
I don't know how to address this one, so I kept the current implementation for now.

Reply:
Fine to leave as is for now. We can decide later if we can merge these traits into one.

Reply:
+1 It would be great if we could reuse that. The problem is that their sample method can't be used to fill a buffer, as ours can. However since we're already asserting in the API that each element is i.i.d., then perhaps we don't actually need that feature?
```rust
    /// Fills the `noise` buffer with elements sampled independently from
    /// the distribution over `F` (i.i.d.) with randomness coming from `seed_stream`.
    fn sample<F: FieldElement, S: SeedStream>(&self, seed_stream: &mut S, noise: &mut [F]);
```

However would also need to implement the `RngCore`, which is slightly more sophisticated than `SeedStream`: https://rust-random.github.io/rand/rand/trait.RngCore.html

I think the trait is stable enough that we are safe to depend on it.

Reply:
Decision on 2023/6/21 call: Take Distribution and RngCore traits.

Reply:
Thanks, I dropped the trait and rebased.